data/transactions/teal.go

```diff
@@ -32,9 +32,14 @@ type EvalDelta struct {
 	GlobalDelta basics.StateDelta `codec:"gd"`
 
 	// When decoding EvalDeltas, the integer key represents an offset into
-	// [txn.Sender, txn.Accounts[0], txn.Accounts[1], ...]
+	// [txn.Sender, txn.Accounts[0], txn.Accounts[1], ..., SharedAccts[0], SharedAccts[1], ...]
 	LocalDeltas map[uint64]basics.StateDelta `codec:"ld,allocbound=config.MaxEvalDeltaAccounts"`
```

Comment:
I see how's data written to `LocalDelta` in `ensureLocalDelta` function but cannot find a test for `LocalDeltas` containing both `Accounts` and  `SharedAccts`.
I also think there should be a helper method resolving LocalDeltas indices to addresses (well maybe just a `ResolveLocalDeltas` that returns addr -> StateDelta list of "tuples". It is also interesting to test/ensure SharedAcct cannot be one of (Sender, Accounts) in order to guarantee canonical indexes (like we had 0 or 1 for addresses that are both Sender and Accounts[0])

Reply:
In `evalStateful_test.go` the tests that check:
```
		require.Equal(t, map[uint64]basics.StateDelta{
			2: {
```
need to use 2 because there is an account in tx.Accounts. However, they don't contain an actual change to that account, so there's no actual use of the `1` index.  I'll add.

Reply:
Helper:
```
func edIndexToAddress(index uint64, txn *transactions.Transaction, shared []basics.Address) string {
	// index into [Sender, txn.Accounts[0], txn.Accounts[1], ..., shared[0], shared[1], ...]
	switch {
	case index == 0:
		return txn.Sender.String()
	case int(index-1) < len(txn.Accounts):
		return txn.Accounts[index-1].String()
	case int(index-1)-len(txn.Accounts) < len(shared):
		return shared[int(index-1)-len(txn.Accounts)].String()
	default:
		return fmt.Sprintf("Invalid Account Index %d in LocalDelta", index)
	}
}
```

Reply:
>That seems like too much rechecking

Agree.

> Helper `edIndexToAddress`...

IMO it makes sense to have this helper near the definition in order to give an example how it works to reader - like a method taking SignedTxnWithAD and listing addresses and their deltas, or move `edIndexToAddress` to there?

Reply:
The trouble is that this method is very "RESTy" - it returns strings rather than addresses. So, while I could move this method into a more core package, and return `basics.Address` from there, then convert to string here (or an error on `nil`), it seems like more code for the small benefit of relocating the function.  The reality is that _only_ REST code needs to interpret the EvalDelta. In the core code, EvalDeltas are "write only".